File: src/types/planner.ts
/** Shared types for the AI Study Plan Generator (Part 3). */

export type StudyTaskType =
  | 'lecture'
  | 'revision'
  | 'practice'
  | 'quiz'
  | 'flashcards'
  | 'assignment'
  | 'break'
  | 'exam-prep'

export type StudyTaskPriority = 'low' | 'medium' | 'high' | 'critical'

export interface StudyTask {
  id: string
  title: string
  subject: string
  type: StudyTaskType
  priority: StudyTaskPriority
  startMinute: number // minutes from midnight
  durationMinutes: number
  day: number // 0 = today, 1 = tomorrow, ... within the visible window
  done: boolean
  aiReason: string
  color: string
}

export interface WeakTopic {
  subject: string
  topic: string
  masteryPct: number
  trend: 'up' | 'down' | 'flat'
}

export interface UpcomingExam {
  id: string
  subject: string
  title: string
  date: string
  daysAway: number
  readiness: number
}

export interface StudyRecommendation {
  id: string
  icon: string
  title: string
  body: string
  actionLabel: string
  kind: 'next' | 'revise' | 'break' | 'lecture' | 'quiz' | 'flashcards' | 'weak-topic'
}

export interface PlannerInputsSummary {
  upcomingExams: number
  weakSubjects: string[]
  strongSubjects: string[]
  availableHoursPerDay: number
  focusScore: number
  quizAccuracy: number
  learningSpeed: 'slower' | 'average' | 'faster'
  currentStreak: number
}

File: backend/app/schemas/planner.py
"""Pydantic schemas for the AI Study Plan Generator API — mirrors src/types/planner.ts."""

from __future__ import annotations

from enum import Enum
from typing import Literal

from pydantic import BaseModel


class StudyTaskOut(BaseModel):
    id: str
    title: str
    subject: str
    type: Literal[
        "lecture", "revision", "practice", "quiz", "flashcards", "assignment", "break", "exam-prep"
    ]
    priority: Literal["low", "medium", "high", "critical"]
    start_minute: int
    duration_minutes: int
    day: int
    done: bool
    ai_reason: str


class UpcomingExamOut(BaseModel):
    id: str
    subject: str
    title: str
    date: str
    days_away: int
    readiness: int


class WeakTopicOut(BaseModel):
    subject: str
    topic: str
    mastery_pct: int
    trend: Literal["up", "down", "flat"]


class StudyRecommendationOut(BaseModel):
    id: str
    icon: str
    title: str
    body: str
    action_label: str
    kind: Literal["next", "revise", "break", "lecture", "quiz", "flashcards", "weak-topic"]


class PlanRegenerationTrigger(str, Enum):
    quiz_completed = "quiz-completed"
    lecture_uploaded = "lecture-uploaded"
    exam_added = "exam-added"
    assignment_submitted = "assignment-submitted"
    performance_improved = "performance-improved"
    performance_declined = "performance-declined"


class RegeneratePlanRequest(BaseModel):
    trigger: PlanRegenerationTrigger
    context: dict = {}


class RegeneratePlanResponse(BaseModel):
    message: str
    tasks: list[StudyTaskOut]
